### PR TITLE
add _formated field to search result

### LINKED ISSUE #75: _formated is not returned anymore
Right now, when transformation are requested on search results (highlight, crop), the transformation are done on the original document, whereas legacy meilisearch used to return both the original document, and a `_formated` version.

There has been talked some time ago to remmove the formated altogether, and avoid the useless sideloading of full documents.

The full documents can still be easily retrieved with the `indexes/:uid/documents/:id` route.

We can decide if we want this feature to stay in transplant.

### PR DESCRIPTION
**Pull request by MarinPostma**
close #75

### CODE REVIEW

File: meilisearch-http/src/index/search.rs [hunk 1]
```diff
@@ -35,10 +37,18 @@ pub struct SearchQuery {
     pub facet_distributions: Option<Vec<String>>,
 }
 
+#[derive(Debug, Clone, Serialize)]
+pub struct SearchHit {
+    #[serde(flatten)]
+    pub document: Document,
+    #[serde(rename = "_formatted", skip_serializing_if = "Map::is_empty")]
```
Comment (Kerollmops) [inline meilisearch-http/src/index/search.rs:44]:
```suggestion
    #[serde(rename = "_formatted", skip_serializing_if = "Document::is_empty")]
```

:)

**Review by curquiza — CHANGES_REQUESTED** [review]
With this request:
```json
{
    "q": "",
    "attributesToRetrieve": ["title", "author"],
    "attributesToHighlight": ["*"]
}
```

I don't get any `_formatted`:
```json
{
    "hits": [
        {
            "author": "Lewis Carroll",
            "title": "Alice In Wonderland"
        },
        {
            "author": "J. R. R. Tolkien",
            "title": "The Hobbit"
        },
        {
            "author": "Jane Austin",
            "title": "Pride and Prejudice"
        },
        {
            "author": "J. K. Rowling",
            "title": "Harry Potter and the Half-Blood Prince"
        },
        {
            "author": "Douglas Adams",
            "title": "The Hitchhiker's Guide to the Galaxy"
        },
        {
            "author": "Antoine de Saint-Exupéry",
            "title": "Le Petit Prince"
        }
    ],
    "nbHits": 6,
    "exhaustiveNbHits": false,
    "query": "",
    "limit": 20,
    "offset": 0,
    "processingTimeMs": 8
}
```

**Review by curquiza — CHANGES_REQUESTED** [review]
I found out two problems:
- the order of the attributes changes for each request (in `_formatted` or not)
- the `displayedAttributes` are not taken into account during the search

**Review by irevoire — APPROVED** [review]

**Comment by irevoire** [comment]
> * the order of the attributes changes for each request (in `_formatted` or not)
Is this really a problem? 
I guess we could use a `BTreeSet` to always return everything in the same order but since we are returning json I don't think we needs to.

File: meilisearch-http/src/index/search.rs [hunk 2]
```diff
@@ -78,19 +88,89 @@ impl Index {
         let mut documents = Vec::new();
         let fields_ids_map = self.fields_ids_map(&rtxn).unwrap();
 
-        let fields_to_display =
-            self.fields_to_display(&rtxn, query.attributes_to_retrieve, &fields_ids_map)?;
+        let displayed_ids: HashSet<FieldId> = fields_ids_map.iter().map(|(id, _)| id).collect();
```
Comment (irevoire) [inline meilisearch-http/src/index/search.rs:91]:
```suggestion
        let displayed_ids = sel.fields_to_display(&rtxn, query.attribute_to_retrieve, &fields_ids_map)?;
        let displayed_ids: HashSet<FieldId> = displayed_ids.iter().map(|(id, _)| id).collect();
```

Comment (irevoire) [reply meilisearch-http/src/index/search.rs:91]: I think this should fix the bug @curquiza found about the `displayedAttributes` not being taken in account during the search

**Review by irevoire — CHANGES_REQUESTED** [review]

**Comment by MarinPostma** [comment]
I agree with @irevoire, this is useless imo, but, since we need to do it anyway, we need to use indexmap

**Comment by curquiza** [comment]
Since we have a criterion where the order of attributes matters, it's important to still return the same order to the user: even if it does not have any impact, it avoids confusion for the users. Until the product team decides the contrary, we need to stay consistent between two searches 🙂

**Comment by MarinPostma** [comment]
@curquiza both issues are now fixed

**Review by curquiza — APPROVED** [review]
Looks good to meeeeeeee 🚀 
well done!!!

**Comment by MarinPostma** [comment]
Bors merge

**Comment by MarinPostma** [comment]
bors merge

**Pull request merged**